deb/ubuntu-bionic/Dockerfile

```diff
@@ -1,30 +1,135 @@
+RUN mkdir -p /build/bin/ && cp /go/src/github.com/*/*/bin/* /build/bin/
+
+# Build the daemon and dependencies
+# TODO: the build scripts from upstream, names the binary with the architecture
+# in its name, and creates a "dockerd" symlink. We don't need those, so either
+# update the upstream scripts, or just execute the right commands ourselves.
+FROM golang-build AS engine-build
+ENV DOCKER_BUILDTAGS="apparmor seccomp selinux"
+ENV RUNC_BUILDTAGS="apparmor seccomp selinux"
+
+# TODO: why are we creating a `tgz` for the sources if we can just add a directory?
+# this would also solve having to do a "mv" below.
+ADD sources/engine.tgz /go/src/github.com/docker/
+RUN mv /go/src/github.com/docker/engine /go/src/github.com/docker/docker
+WORKDIR /go/src/github.com/docker/docker
+RUN PRODUCT=docker ./hack/make.sh dynbinary
+RUN TMP_GOPATH="/go" hack/dockerfile/install/install.sh tini
+RUN TMP_GOPATH="/go" hack/dockerfile/install/install.sh proxy dynamic
+RUN bundles/dynbinary-daemon/dockerd --version
+
+FROM scratch AS completion
+COPY --from=manpages-build /go/src/github.com/docker/cli/contrib/completion /completion
+
+FROM scratch AS manpages
+COPY --from=manpages-build /go/src/github.com/docker/cli/man/man1 /man/man1
+
+FROM scratch AS plugins
+COPY --from=plugins-build /build/bin/. /cli-plugins/
+
+FROM scratch AS cli
+COPY --from=cli-build /usr/local/bin/docker /bin/
+
+FROM scratch AS engine
+COPY --from=engine-build /go/src/github.com/docker/docker/bundles/dynbinary-daemon/. /bin/
+
+# Build this stage with
+# make UBUNTU_VERSIONS=ubuntu-bionic DEBIAN_VERSIONS="" RASPBIAN_VERSIONS="" CLI_DIR=$GOPATH/src/github.com/docker/cli ENGINE_DIR=$GOPATH/src/github.com/docker/docker deb
+#
+# Collect all the artefacts. Note that we didn't build containerd or runc, so
+# we can't run the daemon. We can try the cli against a daemon on the host though;
+# docker run -it --rm -v /var/run/docker.sock:/var/run/docker.sock debbuild-ubuntu-bionic/x86_64
+#
+# This stage is based on build-base, so contains the ~/build-deb directory, which
+# we probably should clean up.
+FROM build-base AS final
+COPY --from=completion  /.      /build/
+COPY --from=manpages    /.      /build/
+COPY --from=plugins     /.      /root/.docker/cli-plugins/
+COPY --from=cli         /bin/.  /usr/local/bin/
+COPY --from=engine      /bin/.  /usr/local/bin/
+RUN echo '{"experimental":"enabled", "debug": true}' > /root/.docker/config.json \
+ && docker --help
+WORKDIR /build/
+
+# Stage to collect all the binaries, which could be used with `--output type=local,dsc=build
+#
+# Build this stage with
+# make UBUNTU_VERSIONS=ubuntu-bionic DEBIAN_VERSIONS="" RASPBIAN_VERSIONS="" CLI_DIR=$GOPATH/src/github.com/docker/cli ENGINE_DIR=$GOPATH/src/github.com/docker/docker binaries
+FROM scratch AS binaries
+COPY --from=completion  /. /
+COPY --from=manpages    /. /
+COPY --from=plugins     /. /
+COPY --from=cli         /. /
+COPY --from=engine      /. /
 
+FROM build-base AS deb
```

Comment:
Haven't tried this stage; the idea will be to;

1. build artefacts
2. copy them to this stage
3. run deb (/rpm) specific steps to package the artefacts as a `.deb` or `.rpm`